rfc/009-distributed-coordinator.md

```diff
@@ -0,0 +1,236 @@
+
+1. Allowing a Coordinator with a different policy to recover is a prerequisite for automatic updates.
+2. The Coordinator policy can't be embedded into the Coordinator during build, and deriving it at runtime is inconvenient.
+
+Thus, we introduce a new field to the manifest that specifies roles available to the verified identity.
+For now, the only known role is `coordinator`, but this could be extended in the future (for example: delegate CAs).
+
+```golang
+type PolicyEntry struct {
+  SANs             []string
+  WorkloadSecretID string `json:",omitempty"`
+  Roles            []string `json:",omitempty"` // <-- New in RFC009.
+}
+```
+
+During `generate`, we append the value of the `contrast.edgeless.systems/pod-role` annotation to the policy entry.
+If there is no coordinator among the resources, we add a policy entry for the embedded coordinator policy.
+
+### Peer recovery
+
+The peer recovery process is attempted by Coordinators that are in [Recovery mode](#recovery-mode):
+
+1. Once after startup, to speed up initialization.
+2. Once as callback to the `Watch` event, .
+3. Periodically from a goroutine, to reconcile missed watch events.
+
+The process starts with a DNS request for the `coordinator-peers` name to discover ready Coordinators.
+For each of the ready Coordinators, the recovering Coordinator calls a new `meshapi.Recover` method.
+
+```proto
+service MeshAPI {
+  rpc NewMeshCert(NewMeshCertRequest) returns (NewMeshCertResponse);
+  rpc Recover(RecoverRequest) returns (RecoverResponse); // <-- New in RFC009.
+}
+
+message RecoverRequest {}
+
+message RecoverResponse {
+  bytes Seed = 1;
+  bytes Salt = 2;
+  bytes RootCAKey = 3;
+  bytes RootCACert = 4;
+  bytes MeshCAKey = 5;
+  bytes MeshCACert = 6;
+  bytes LatestManifest = 7;
+}
+```
+
+When this method is called, the serving Coordinator:
+
+1. Provides an attestation statement to the client (this isn't the case for the mesh API right now).
+2. Queries the current state and attaches it to the context (this already happens automatically for all `meshapi` calls).
+3. Verifies that the client identity is allowed to recover by the state's manifest (has the `coordinator` role).
+4. Fills the response with values from the current state.
+
+The complete flow from the recovering Coordinator's perspective:
+
+1. Load the latest manifest from persistence without verifying the signature.
+2. Build an aTLS client from the unverified manifest:
+   * Populate ValidateOpts with reference values.
+   * Add a callback that verifies the policy hash with the roles from the unverified manifest.
+   * Add standard client attestation.
+3. Call the serving Coordinator's `Recover` endpoint.
+4. Compare the received manifest with the temporary manifest.
+5. Set up a temporary `SeedEngine` with the received parameters.
+6. Load the manifest history from persistence, verifying the signature with the temporary seed engine.
+7. Compare the loaded manifest with the received manifest.
```

Comment:
Just to clarify, this means if a coordinator was in recovery state while a manifest was updated, it can't be recovered anymore, right?

Reply:
If the comparison fails, the recovering Coordinator should start over at (1), so it should eventually succeed.